sdks/go/pkg/beam/io/synthetic/source.go

```diff
@@ -0,0 +1,151 @@
+//
+// The syntheticSourceFn is expected to receive elements of type sourceConfig
+// and follow that config to determine its behavior when splitting and emitting
+// elements.
+type sourceFn struct {
+	rng *rand.Rand
+}
+
+// CreateInitialRestriction creates an offset range restriction representing
+// the number of elements to emit.
+func (fn *sourceFn) CreateInitialRestriction(config SourceConfig) offsetrange.Restriction {
+	return offsetrange.Restriction{
+		Start: 0,
+		End:   int64(config.NumElements),
+	}
+}
+
+// SplitRestriction splits restrictions equally according to the number of
+// initial splits specified in SourceConfig. Each restriction output by this
+// method will contain at least one element, so the number of splits will not
+// exceed the number of elements.
+func (fn *sourceFn) SplitRestriction(config SourceConfig, rest offsetrange.Restriction) (splits []offsetrange.Restriction) {
+	if config.InitialSplits <= 1 {
+		// Don't split, just return original restriction.
+		splits = append(splits, rest)
+		return splits
+	}
+
+	num := int64(config.InitialSplits)
+	offset := rest.Start
+	size := rest.End - rest.Start
+	for i := int64(0); i < num; i++ {
+		split := offsetrange.Restriction{
+			Start: offset + (i * size / num),
+			End:   offset + ((i + 1) * size / num),
+		}
+		// Skip restrictions that end up empty.
+		if split.End-split.Start <= 0 {
+			continue
+		}
+		splits = append(splits, split)
+	}
+	return splits
+}
+
+// RestrictionSize outputs the size of the restriction as the number of elements
+// that restriction will output.
+func (fn *sourceFn) RestrictionSize(config SourceConfig, rest offsetrange.Restriction) float64 {
+	return float64(rest.End - rest.Start)
+}
+
+// CreateTracker just creates an offset range restriction tracker for the
+// restriction.
+func (fn *sourceFn) CreateTracker(rest offsetrange.Restriction) *offsetrange.Tracker {
+	return offsetrange.NewTracker(rest)
+}
+
+// StartBundle sets up the random number generator.
+func (fn *sourceFn) StartBundle() {
+	fn.rng = rand.New(rand.NewSource(time.Now().UnixNano()))
+}
+
+// ProcessElement creates a number of random elements based on the restriction
+// tracker received. Each element is a random byte slice key and value, in the
+// form of KV<[]byte, []byte>,
+func (fn *sourceFn) ProcessElement(rt *offsetrange.Tracker, config SourceConfig, emit func([]byte, []byte)) error {
+	for i := rt.Rest.Start; rt.TryClaim(i) == true; i++ {
+		key := make([]byte, 8)
+		val := make([]byte, 8)
+		if _, err := fn.rng.Read(key); err != nil {
+			return err
+		}
+		if _, err := fn.rng.Read(val); err != nil {
+			return err
+		}
+		emit(key, val)
+	}
+	return nil
+}
+
+// DefaultSourceConfig creates a SourceConfig with intended defaults for its
+// fields. SourceConfigs should be initialized with this method.
+func DefaultSourceConfig() SourceConfig {
+	return SourceConfig{
+		NumElements:   0, // Number of elements for the synthetic source to produce.
```

Comment:
The comments on the fields here seem to  be redundant to the ones on the definition struct fields.

Consider switching them with the *why*s for those default values.

There's an idiom in go to "make the zero value useful". In this case the zero value is SourceConfig{}. WRT the implementation, this is identical to what's being provided by the DefaultSourceConfig() method since InitialSplits is "default 1" anyway. 

I do like the declaration that DefaultSourceConfig() is recommended for how to configure initially since it'll set th defaults, especially when new fields are being added, so behavior doesn't unnecessarily change when the new feature is added.

Reply:
Agreed. I think I added these first when commenting stuff and didn't erase them when I added the more descriptive comments. Done.